### ISSUE TITLE
Include array index as a final argument to createArraySelector callback

### ISSUE THREAD

**Issue by pbadenski** [heyimalex/reselect-map#20]
In order to make this possible:

```
const mul = createArraySelector(
  numbers,
  (number) => [index, number]
);
```

**Comment by heyimalex** [comment]
This was intentional; memoization is bound to a "key" and array index doesn't make sense as a key because re-orders trigger `length - index` total changes. Changing this would opt everyone into that behavior, since it makes the mapping function dependent on the index and the mapping functions need to be pure. [Maybe some more discussion about this choice here.](https://github.com/heyimalex/reselect-map/issues/4)

**Comment by pbadenski** [comment]
Makes sense. Although it does optimise towards reselect-map in computation use case (which what this library was created for, so that makes sense 😆). I'm plugging in reselect-map for inputArray -> outputArray transformation to ensure output array only changes at particular indices where inputArray also changed and other elements remain unchanged. The behaviour with ordering is actually what I would be expecting.

Hope you can help... What would be the easiest way for me to roll out the custom implementation? Would it be possible to expose a bit more of the API (eg. `createMappedSelectorCreator` and `memoizeMap`)?

**Comment by heyimalex** [comment]
Yeah, we can definitely expose those. I think just exporting and then adding to the typescript definitions is all that needs to be done if you want to submit a pr, or I can try and tackle it sometime this week.

**Comment by pbadenski** [comment]
cool, I'll have a go at it, thanks!
